repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +88,57 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      def stmtState = _statements(statementId).state
+
+      if (stmtState != StatementState.Cancelled) {
+        lock.synchronized {
+          if (stmtState != StatementState.Cancelled) {
+            _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+            _sc.foreach(
+              _.setJobGroup(statementId.toString, s"Job group for statement $statementId"))
+          }
+        }
+      }
+
+      val executeResult = if (stmtState != StatementState.Cancelled) {
+        executeCode(statementId, code)
+      } else {
+        null
+      }
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      if (stmtState != StatementState.Cancelled) {
+        lock.synchronized {
+          if (stmtState != StatementState.Cancelled) {
+            _statements(statementId) =
+              new Statement(statementId, StatementState.Available, executeResult)
+          }
+        }
+      }
     }
+
     statementId
   }
 
+  def cancel(statementId: Int): Unit = {
```

Comment:
It is hard to distinguish `cancelling` and `cancelled`:

1. `sc.cancelJobGroup` is non-blocking, it will return immediately.
2. `sc.cancelJobGroup` is asynchronous, Job will be cancelled in future after this is called.
3. Caller doesn't know when job is cancelled, so it doesn't know when to change the statement to `Cancelled`.

It is hard to set `cancelling` and `cancelled` in Livy side.